### ISSUE TITLE
Valid RFC-3164 syslog date/time fails to parse due to erroneous DST gap

### ISSUE THREAD

**Issue by neirbowj** [logstash-plugins/logstash-filter-date#96]
An [RFC-3164 TIMESTAMP](https://tools.ietf.org/html/rfc3164#section-4.1.2) in the interval between 2am and 3am local time on April 2 will fail to parse. This suggests that there is a DST gap at that time. However, the UTC offset of the timestamps on either side of the gap indicate otherwise and, moreover, there shouldn't be a gap at that time in this time zone. If a year is specified, the parse succeeds.

- Version: logstash 5.3.0, logstash-filter-date 3.1.3
- Operating System: FreeBSD 10.3-STABLE (r302062)
- Config File:

```
input { stdin { } }
filter { date { match => [
    "message",
    "MMM d HH:mm:ss",
    "MMM d HH:mm:ss yyyy"
] } }
output { stdout { codec => rubydebug } }
```
Also, the system is configured for America/New_York.

```
$ sha1 -r /etc/localtime /usr/share/zoneinfo/America/New_York
6e43d0b5a46ed5ba78da5c7e9dcf319b27d769e7 /etc/localtime
6e43d0b5a46ed5ba78da5c7e9dcf319b27d769e7 /usr/share/zoneinfo/America/New_York
```

- Sample Data:
```
Apr 2 01:59:59
Apr 2 02:00:00
Apr 2 02:59:59
Apr 2 03:00:00
Apr 2 01:59:59 2017
Apr 2 02:00:00 2017
Apr 2 02:59:59 2017
Apr 2 03:00:00 2017
```
- Steps to Reproduce:

```
$ cat sample | sudo /usr/local/logstash/bin/logstash -f datebug.conf
[...]
{
    "@timestamp" => 2017-04-02T05:59:59.000Z,
      "@version" => "1",
          "host" => "host1",
       "message" => "Apr 2 01:59:59"
}
{
    "@timestamp" => 2017-05-07T19:45:55.110Z,
      "@version" => "1",
          "host" => "host1",
       "message" => "Apr 2 02:00:00",
          "tags" => [
        [0] "_dateparsefailure"
    ]
}
{
    "@timestamp" => 2017-05-07T19:45:55.111Z,
      "@version" => "1",
          "host" => "host1",
       "message" => "Apr 2 02:59:59",
          "tags" => [
        [0] "_dateparsefailure"
    ]
}
{
    "@timestamp" => 2017-04-02T07:00:00.000Z,
      "@version" => "1",
          "host" => "host1",
       "message" => "Apr 2 03:00:00"
}
{
    "@timestamp" => 2017-04-02T05:59:59.000Z,
      "@version" => "1",
          "host" => "host1",
       "message" => "Apr 2 01:59:59 2017"
}
{
    "@timestamp" => 2017-04-02T06:00:00.000Z,
      "@version" => "1",
          "host" => "host1",
       "message" => "Apr 2 02:00:00 2017"
}
{
    "@timestamp" => 2017-04-02T06:59:59.000Z,
      "@version" => "1",
          "host" => "host1",
       "message" => "Apr 2 02:59:59 2017"
}
{
    "@timestamp" => 2017-04-02T07:00:00.000Z,
      "@version" => "1",
          "host" => "host1",
       "message" => "Apr 2 03:00:00 2017"
}
```

Here is the output from the same config and sample data presented to logstash 5.0.2 with logstash-filter-date 3.0.3.

```
{
    "@timestamp" => 2017-04-02T05:59:59.000Z,
      "@version" => "1",
          "host" => "host2",
       "message" => "Apr 2 01:59:59",
          "tags" => []
}
{
    "@timestamp" => 2017-04-02T06:00:00.000Z,
      "@version" => "1",
          "host" => "host2",
       "message" => "Apr 2 02:00:00",
          "tags" => []
}
{
    "@timestamp" => 2017-04-02T06:59:59.000Z,
      "@version" => "1",
          "host" => "host2",
       "message" => "Apr 2 02:59:59",
          "tags" => []
}
{
    "@timestamp" => 2017-04-02T07:00:00.000Z,
      "@version" => "1",
          "host" => "host2",
       "message" => "Apr 2 03:00:00",
          "tags" => []
}
{
    "@timestamp" => 2017-04-02T05:59:59.000Z,
      "@version" => "1",
          "host" => "host2",
       "message" => "Apr 2 01:59:59 2017",
          "tags" => []
}
{
    "@timestamp" => 2017-04-02T06:00:00.000Z,
      "@version" => "1",
          "host" => "host2",
       "message" => "Apr 2 02:00:00 2017",
          "tags" => []
}
{
    "@timestamp" => 2017-04-02T06:59:59.000Z,
      "@version" => "1",
          "host" => "host2",
       "message" => "Apr 2 02:59:59 2017",
          "tags" => []
}
{
    "@timestamp" => 2017-04-02T07:00:00.000Z,
      "@version" => "1",
          "host" => "host2",
       "message" => "Apr 2 03:00:00 2017",
          "tags" => []
}
```

**Comment by guyboertje** [comment]
@neirbowj - This will be fixed in v 3.1.6.
The problem was with when a year was not in the pattern. We do some year guessing to determine what year a time belongs to. This was using a default year of 2000 - and surprise, DST change in 2000 for NYC was April 2.

**Comment by neirbowj** [comment]
Ah, in bc7918f, I see. Thank you.
